app/controllers/dashboard_controller.rb

```diff
@@ -2,54 +2,45 @@ class DashboardController < ApplicationController
   before_action :authenticate_user!
   before_action :logged_in_user
   include DashboardHelper
+
   def index
     today = date_to_url(DateTime.now)
     redirect_to dashboard_path(today)
   end
   def show
     @chosen_date = params[:id].to_date
-    @query_dates = Income.select(:date).distinct.limit(5).order('date DESC').reverse
-    @dates = Array.new()
-    @query_dates.each do |item|
-      @dates << { active: is_chosen_date?(item.date, @chosen_date), 
-                  date_name: name_of_day(item.date),
-                  date_value: item.date.to_s(:db) 
-                }
-    end
-    @general_expenses = GeneralExpense.list.where(date: @chosen_date).decorate
-    @incomes_truck_routes = Income.truck_routes.where(date: @chosen_date).decorate
-    @incomes_brand_shops = Income.brand_shops.where(date: @chosen_date).decorate
-    @incomes_external_shops = Income.external_shops.where(date: @chosen_date).decorate
+    last_workdays_data(@chosen_date)
+    fetch_data(@chosen_date)
     @chosen_date = @chosen_date.strftime("%Y-%m-%d, %A")
     @general_expense = GeneralExpense.new
-    @income = Income.new
-    @external_shops = ExternalShop.all.map{|c| [ c.name, c.id ]}
-    @drivers = Driver.all.map{|c| [ "#{c.name} #{c.surname}", c.id ]}
   end
   def create
     if is_workday_created?(params[:date])
       flash[:error] = "Workday already exists."
     else
-      weekday = params[:date].to_date.wday
-      truck_routes = TruckRoute.where("weekdays LIKE ?", "%#{weekday}%")
-      truck_routes.each do |shop|
-        shop.incomes.create(date: params[:date])
-      end     
-      brand_shops = BrandShop.all
-      brand_shops.each do |shop|
-        shop.incomes.create(date: params[:date])
-      end
-      external_shops = ExternalShop.where(daily_payments: true)
-      external_shops.each do |shop|
-        shop.incomes.create(date: params[:date])
-      end
+      create_workday_elements(params[:date])
       flash[:success] = "New workday created successfully."
     end
     redirect_to dashboard_path(date_to_url(params[:date]))
   end
 
   private
-
+    def last_workdays_data(day)
+      query_dates = Income.last_workdays
+      @dates = Array.new()
+      query_dates.each do |item|
+        @dates << { active: is_chosen_date?(item.date, day), 
+                    date_name: name_of_day(item.date),
+                    date_value: item.date.to_s(:db) }
+      end
+    end
+    def fetch_data(day)
+      @general_expenses = GeneralExpense.day_list(day).decorate
+      @incomes_truck_routes = Income.truck_routes(day).decorate
+      @incomes_brand_shops = Income.brand_shops(day).decorate
+      @incomes_external_shops = Income.external_shops(day).decorate
+      @drivers = Driver.all.map{|c| [ "#{c.name} #{c.surname}", c.id ]}
+    end
```

Comment:
it's good that you have splitted one huge method into few smaller ones :+1:  but as you can see now, code has been encapsulated into private methods (good practice says "dont test private methods"). What you can do here is extract the code from a controller into some other objects. Someone called them `presenters` or `view_objects`... either way the point is to have a class which looks like:
```ruby
class DashboardPresenter
  def incomes_truck_routes
    #...
  end

  def incomes_brand_shops
   # ...
  end
end
```
this kind of plain ruby object is easy to test because it doesnt need a controller. Your controller, by the way, will become much cleaner working with presenters instead of doing all the job in the private methods. Does it make sense to you?